src/QsCompiler/TextProcessor/CompletionParsing.fs

```diff
@@ -5,49 +5,110 @@
 /// closed if it is still valid for the bracket to be closed later.
 module Microsoft.Quantum.QsCompiler.TextProcessing.CompletionParsing
 
+#nowarn "40"
+
 open System
 open FParsec
+open Microsoft.Quantum.QsCompiler.TextProcessing.ExpressionParsing
 open Microsoft.Quantum.QsCompiler.TextProcessing.Keywords
 open Microsoft.Quantum.QsCompiler.TextProcessing.ParsingPrimitives
 open Microsoft.Quantum.QsCompiler.TextProcessing.SyntaxBuilder
-open Microsoft.Quantum.QsCompiler.TextProcessing.TypeParsing
 
 
 /// Describes the environment of a code fragment in terms of what kind of completions are available.
 type CompletionEnvironment =
     /// The code fragment is inside a namespace but outside any callable.
     | NamespaceTopLevel
+    /// The code fragment is a statement inside a callable.
+    | Statement
 
 /// Describes the kind of identifier that is expected at a particular position in the source code.
 type IdentifierKind =
-    /// The identifier is a new symbol declaration.
-    | Declaration
-    /// The identifier is a type name.
-    | Type
-    /// The identifier is a characteristic set name.
-    | Characteristic
     /// The identifier is the given keyword.
     | Keyword of string
+    /// The identifier is a variable or callable name.
+    | Variable
+    /// The identifier is a user-defined type name.
+    | UserDefinedType
+    /// The identifier is a type parameter.
+    | TypeParameter
+    /// The identifier is a new symbol declaration.
+    | Declaration
     /// The identifier is a namespace.
     | Namespace
     /// The identifier is a member of the given namespace and has the given kind.
     | Member of string * IdentifierKind
+    /// The identifier is a named item in a user-defined type.
+    // TODO: Add information so completion knows the type being accessed.
+    | NamedItem
+
+/// The exception that is thrown when the completion parser can't parse the text of a fragment.
+exception CompletionParserError of detail : string * error : ParserError with
```

Comment:
The exception is basically to make it slightly easier to use from C#, so instead of checking a DU I can catch an exception. I could have `GetExpectedIdentifiers` directly return the `ParserResult` and check Success/Failure from C# if you think that'd be better.

Reply:
Ah, I see. I would like to avoid having any direct references to FParsed outside the TextProcessor. I ususally would do a top-level routine in the parser that is meant to be called from C# opposed to calling a parser directly, and that could then e.g. return a QsNullable. This to me is a borderline case regarding whether the execeptions you are throwing are really unexpected failures, or if you are using them as control flow.

Reply:
`GetExpectedIdentifiers` now returns `CompletionResult` (a custom DU type) instead of throwing an exception.